server-mirror-git/src/main/java/com/linecorp/centraldogma/server/internal/mirror/SshGitMirror.java

```diff
@@ -0,0 +1,257 @@
+            final URIish uri = git.remoteUri();
+            int port = uri.getPort();
+            if (port <= 0) {
+                port = 22; // Use the SSH default port it unspecified.
+            }
+            logger.trace("Connecting to {}:{}", uri.getHost(), port);
+            final ClientSession session = sshClient.connect(uri.getUser(), uri.getHost(), port)
+                                                   .verify(GitModuleProperties.CONNECT_TIMEOUT.getRequired(
+                                                           sshClient))
+                                                   .getSession();
+            session.auth().verify(GitModuleProperties.AUTH_TIMEOUT.getRequired(session));
+            logger.trace("The session established: {}", session);
+            return session;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void configureCredential(SshClient client) {
+        final MirrorCredential c = credential();
+        if (c instanceof PasswordMirrorCredential) {
+            client.setFilePasswordProvider(passwordProvider(((PasswordMirrorCredential) c).password()));
+        } else if (c instanceof PublicKeyMirrorCredential) {
+            final PublicKeyMirrorCredential cred = (PublicKeyMirrorCredential) credential();
+            final Collection<KeyPair> keyPairs;
+            try {
+                keyPairs = keyPairResourceParser.loadKeyPairs(null, NamedResource.ofName(cred.username()),
+                                                              passwordProvider(cred.passphrase()),
+                                                              cred.privateKey());
+                client.setKeyIdentityProvider(KeyIdentityProvider.wrapKeyPairs(keyPairs));
+            } catch (IOException | GeneralSecurityException e) {
+                throw new MirrorException("Unexpected exception while loading private key. username: " +
+                                          cred.username() + ", publicKey: " +
+                                          publicKeyPreview(cred.publicKey()), e);
+            }
+        }
+    }
+
+    private static FilePasswordProvider passwordProvider(@Nullable String passphrase) {
+        if (passphrase == null) {
+            return FilePasswordProvider.EMPTY;
+        }
+
+        return FilePasswordProvider.of(passphrase);
+    }
+
+    private static Consumer<TransportCommand<?, ?>> transportCommandConfigurator(
+            GitSshdSessionFactory sessionFactory) {
+        return command -> {
+            // Need to set the credentials provider, otherwise NPE is raised when creating GitSshdSession.
+            command.setCredentialsProvider(NoopCredentialsProvider.INSTANCE);
+            command.setTransportConfigCallback(transport -> {
+                final SshTransport sshTransport = (SshTransport) transport;
+                sshTransport.setSshSessionFactory(sessionFactory);
+            });
+        };
+    }
```

Comment:
How about moving this method to `DefaultGitSshdSessionFactory` which may be simpler?
```java
class DefaultGitSshdSessionFactory {
  ...

  void configureCommand(TransportCommand<?, ?> command) {
    // Need to set the credentials provider, otherwise NPE is raised when creating GitSshdSession.
    command.setCredentialsProvider(NoopCredentialsProvider.INSTANCE);
    command.setTransportConfigCallback(transport -> {
        final SshTransport sshTransport = (SshTransport) transport;
        sshTransport.setSshSessionFactory(this);
    });
  }
}
```

Reply:
That's a good suggestion. Addressed. 😉